asserts/model.go

```diff
@@ -283,9 +318,74 @@ func checkModelSnap(snap map[string]interface{}, grade ModelGrade, modelIsClassi
 		DefaultChannel: defaultChannel,
 		Presence:       presence, // can be empty
 		Classic:        isClassic,
+		Components:     components, // can be empty
 	}, nil
 }
 
+func checkComponentsForMaps(m map[string]interface{}, validModes []string, what string) (map[string]ModelComponent, error) {
+	const compsField = "components"
+	value, ok := m[compsField]
+	if !ok {
+		return nil, nil
+	}
+	comps, ok := value.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("%q %s must be a map from strings to components",
+			compsField, what)
+	}
+
+	res := make(map[string]ModelComponent, len(comps))
+	for name, comp := range comps {
+		// Name of component follows the same rules as snap components
+		if err := naming.ValidateSnap(name); err != nil {
+			return nil, fmt.Errorf("invalid component name %s", name)
+		}
+
+		// "comp: required|optional" case
+		presence, ok := comp.(string)
+		if ok {
+			if !strutil.ListContains(validSnapPresences, presence) {
+				return nil, fmt.Errorf("presence of component %q %s must be one of required|optional", name, what)
+			}
+			res[name] = ModelComponent{Presence: presence, Modes: nil}
+			continue
+		}
+
+		// try map otherwise
+		compFields, ok := comp.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("%q component %s must be a map of strings to components",
+				name, what)
+		}
+		// Error out if unexpected entry
+		for key := range compFields {
+			if !strutil.ListContains([]string{"presence", "modes"}, key) {
+				return nil, fmt.Errorf("entry %q of component %s %s is unknown", key, name, what)
+			}
+		}
+		presence, err := checkOptionalStringWhat(compFields, "presence", what)
+		if err != nil {
+			return nil, err
+		}
+		if !strutil.ListContains(validSnapPresences, presence) {
+			return nil, fmt.Errorf("presence of component %q %s must be one of required|optional", name, what)
+		}
+		modes, err := checkStringListInMap(compFields, "modes",
+			fmt.Sprintf("%s %s %s", "modes of component", name, what), validSnapMode)
+		if err != nil {
+			return nil, err
+		}
+		for _, m := range modes {
```

Comment:
should we do?
```
if len(modes) == 0 {
    modes = validModes
}
```

Reply:
I believe, by default, the component should inherit the modes from its snap.

Reply:
Right, I've fixed this now